lib/matplotlib/pyplot.py

```diff
@@ -1479,10 +1563,19 @@ def subplots(nrows=1, ncols=1, *, sharex=False, sharey=False, squeeze=True,
     return fig, axs
 
 
-def subplot_mosaic(mosaic, *, sharex=False, sharey=False,
-                   width_ratios=None, height_ratios=None, empty_sentinel='.',
-                   subplot_kw=None, gridspec_kw=None,
-                   per_subplot_kw=None, **fig_kw):
+def subplot_mosaic(
+    mosaic: str | list[list[Any]],
```

Comment:
According to https://stackoverflow.com/a/53845083, you should be able to annotate the recursive list now (the linked mypy issue there is fixed.)

Reply:
I'll try it out, though I suspect it may be harder to parse as a human than just seeing `Any` plus the docstring (and will require some additional named types, since you can't do a forward reference inline)

Reply:
It does in fact work, and isn't as hard to read as I feared:

```python
    HashableList = list[Hashable | "HashableList"]
    def subplot_mosaic(
        self,
        mosaic: str | list[HashableList],
```

(tested in the `figure.pyi` version of this function)

It does require the stringified forward reference, which is otherwise mostly not needed in `pyi` files or `py` files with the `__future__` import, so there's a mild confusion potential from that, but not actually that bad.

Holding off until committing this as `HashableList` is something that might go in the `mpl.typing` module as it is purely for typing purposes.